test/array.cpp

```diff
@@ -0,0 +1,400 @@
+   auto b_it = b.begin();
+
+   bool passed = *a_it++ == 1 &&
+                 *a_it++ == 8 &&
+                 *b_it++ == 1 &&
+                 *b_it++ == 8;
+
+   a_it = a.begin();
+   *a_it = 4;
+   b_it = b.begin();
+
+   return passed &&
+          *a_it++ == 4 &&
+          *a_it++ == 8 &&
+          *b_it++ == 4 &&
+          *b_it++ == 8;
+} CAMP_TEST_END(array, begin)
+
+CAMP_TEST_BEGIN(array, cbegin)
+{
+   camp::array<int, 2> a = {1, 8};
+   auto a_it = a.cbegin();
+
+   const camp::array<int, 2>& b{a};
+   auto b_it = b.cbegin();
+
+   return *(a_it++) == 1 &&
+          *(a_it++) == 8 &&
+          *(b_it++) == 1 &&
+          *(b_it++) == 8;
+} CAMP_TEST_END(array, cbegin)
+
+CAMP_TEST_BEGIN(array, end)
+{
+   camp::array<int, 2> a = {1, 8};
+   auto a_it = a.end();
+
+   const camp::array<int, 2>& b{a};
+   auto b_it = b.end();
+
+   bool passed = *(--a_it) == 8 &&
+                 *(--a_it) == 1 &&
+                 *(--b_it) == 8 &&
+                 *(--b_it) == 1;
+
+   a_it = a.end();
+   *(--a_it) = 4;
+
+   a_it = a.end();
+   b_it = b.end();
+
+   return passed &&
+          *(--a_it) == 4 &&
+          *(--a_it) == 1 &&
+          *(--b_it) == 4 &&
+          *(--b_it) == 1;
+} CAMP_TEST_END(array, end)
+
+CAMP_TEST_BEGIN(array, cend)
+{
+   camp::array<int, 2> a = {1, 8};
+   auto a_it = a.cend();
+   --a_it;
+
+   const camp::array<int, 2>& b{a};
+   auto b_it = b.cend();
+
+   return *(a_it--) == 8 &&
+          *(a_it--) == 1 &&
+          *(--b_it) == 8 &&
+          *(--b_it) == 1;
+} CAMP_TEST_END(array, cend)
+
+CAMP_TEST_BEGIN(array, empty)
+{
+   // Zero sized arrays are technically not allowed,
+   // and are explicitly disallowed in device code.
+   camp::array<double, 1> a{1.0};
+
+   return !a.empty();
+} CAMP_TEST_END(array, empty)
+
+CAMP_TEST_BEGIN(array, size)
+{
+   // Zero sized arrays are technically not allowed,
+   // and are explicitly disallowed in device code.
+   camp::array<double, 2> a{1.0, 3.0};
+
+   return a.size() == 2;
+} CAMP_TEST_END(array, size)
+
+CAMP_TEST_BEGIN(array, max_size)
+{
+   // Zero sized arrays are technically not allowed,
+   // and are explicitly disallowed in device code.
+   camp::array<double, 2> a{1.0, 3.0};
+
+   return a.size() == 2;
+} CAMP_TEST_END(array, max_size)
+
+CAMP_TEST_BEGIN(array, fill)
+{
+   camp::array<int, 3> a{1, 2, 3};
+   a.fill(0);
+
+   return a[0] == 0 &&
+          a[1] == 0 &&
+          a[2] == 0;
+} CAMP_TEST_END(array, fill)
+
+CAMP_TEST_BEGIN(array, equal)
+{
+   camp::array<int, 2> a{1, 2};
+   camp::array<int, 2> b{1, 3};
+
+   return a == a &&
+          !(a == b);
+} CAMP_TEST_END(array, equal)
+
+CAMP_TEST_BEGIN(array, not_equal)
+{
+   camp::array<int, 2> a{1, 2};
+   camp::array<int, 2> b{1, 3};
+
+   return a != b &&
+          !(a != a);
+} CAMP_TEST_END(array, not_equal)
+
+CAMP_TEST_BEGIN(array, less_than)
+{
+   camp::array<int, 2> a{1, 2};
+   camp::array<int, 2> b{1, 3};
+
+   return a < b &&
+          !(b < a);
```

Comment:
Should test the case when they are equals as well.
```suggestion
   return !(a < a) &&
          a < b &&
          !(b < a);
```